Enable capsule updates

SergiiDmytruk opened the pull request:
This lacks FMP for update to actually do anything.

***

New CAPSULE_ENABLE define in DasharoPayloadPkg.dsc adds required libraries/DXEs/PCDs and enables building of CapsuleApp which is useful for testing.

BlSupportPei now switches boot mode to BOOT_ON_FLASH_UPDATE if there are any capsules.  It's not guarded by any PCD under the assumption that coreboot won't pass capsules if EDK shouldn't be handling them.

EsrtDxe is enabled to manage and cache ESRT entries (caching is supposedly necessary after a capsule is processed, although don't really understand why).  ProcessCapsules() internally looks up EsrtManagementProtocol and calls SyncEsrtFmp() to import data from available FMP instances.

PlatformBootManagerLib was made to call ProcessCapsules() (twice).

PatrickRudolph committed:
UefiCpuPkg: Disable MTRR programming for DasharoPayloadPkg

The MTRRs have already been programmed by FSB (coreboot).

Change-Id: I8c6e7657b789edac8ee83a89318209a598712db8
Signed-off-by: Patrick Rudolph <patrick.rudolph@9elements.com>
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
DasharoPayloadPkg/BlSupportDxe: Publish ESRT with an entry for coreboot

Use firmware information passed by coreboot in CBMEM to make a
single-element ESRT.  Later on ESRT will likely be constructed by
EsrtDxe from data provided by FMP instances, but until then this
should be good enough for viewing this information in an OS and posting
of capsules.

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
DasharoPayloadPkg/Coreboot.h: Update value of CB_TAG_CAPSULE

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
MdeModulePkg: Add PcdCapsuleFmpSupport to set EFI_OS_INDICATIONS_FMP_CAPSULE_SUPPORTED

Don't know why EDK never sets this bit (there is only definition with no
uses), maybe nothing actually checks its value.  Whatever the reason is,
it seems like a good idea to make setting it possible.

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
DasharoPayloadPkg: Optionally enable update capsules

New CAPSULE_SUPPORT define in DasharoPayloadPkg.dsc adds required
libraries/DXEs/PCDs and enables building of CapsuleApp which is useful
for testing.

BlSupportPei now switches boot mode to BOOT_ON_FLASH_UPDATE if there are
any capsules.  It's not guarded by any PCD under the assumption that
coreboot won't pass capsules if EDK shouldn't be handling them.

EsrtDxe is enabled to manage and cache ESRT entries (caching is
supposedly necessary after a capsule is processed, although don't really
understand why).  ProcessCapsules() internally looks up
EsrtManagementProtocol and calls SyncEsrtFmp() to import data from
available FMP instances.

PlatformBootManagerLib was made to call ProcessCapsules() (twice).

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

krystian-hebel commented on DasharoPayloadPkg/Library/PlatformBootManagerLib/PlatformBootManager.c:
```diff
@@ -1567,6 +1582,19 @@ PlatformBootManagerAfterConsole (
+    Status = ProcessCapsules ();
+    if (EFI_ERROR (Status)) {
+      DEBUG((DEBUG_ERROR, "%a(): ProcessCapsule() failed with: %r\n", __FUNCTION__, Status));
+    }
```
We have a requirement for not allowing to boot to OS with SMI handler enabled, this seems to be a good place to enforce it.

```suggestion
    }
    // Reset system to disable SMI handler
    gRT->ResetSystem (EfiResetCold, EFI_SUCCESS, 0, NULL);
```

SergiiDmytruk replied:
https://github.com/Dasharo/edk2/compare/e4bb5bf01ba0efbe91fe702bfb08dd9d47b9f174..4fd97ed77cc98d27a2465d7b5898daa8908f0028

SergiiDmytruk committed:
DasharoPayloadPkg/CbParseLib: Add CB_TAG_FW_INFO parsing API

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

krystian-hebel commented on DasharoPayloadPkg/DasharoPayloadPkg.dsc:
```diff
@@ -109,6 +109,7 @@
   DEFINE PERFORMANCE_MEASUREMENT_ENABLE = FALSE
   DEFINE RAM_DISK_ENABLE                = FALSE
   DEFINE APU_CONFIG_ENABLE              = FALSE
+  DEFINE CAPSULE_ENABLE                 = FALSE
```
Can this be passed from coreboot? I tried using `BUILD_STR += -D CAPSULE_ENABLE=TRUE`, but it doesn't change the value. I think this may be broken in the same way as `DASHARO_SYSTEM_FEATURES_ENABLE` wasn't properly written to `PcdShowMenu`. It seems to work properly in FDF, but in DSC all `!if $(CAPSULE_ENABLE) == TRUE` (at least in `LibraryClasses` section) are not taken, regardless of what value has been passed by coreboot.

SergiiDmytruk replied:
I tested it with `BUILD_STR += -D CAPSULE_ENABLE=TRUE` and `BUILD_STR += -D CAPSULE_ENABLE=FALSE` in coreboot.  Are you sure you're triggering a rebuild of EDK?  I usually do `rm -f build/UEFIPAYLOAD.fd build/coreboot.rom` before `make` to force rebuild of EDK.

krystian-hebel replied:
Never mind, it was my fault after all. I was appending to `BUILD_STR` based on `CONFIG_DRIVERS_EFI_UPDATE_CAPSULES`, but I forgot to pass it in `payloads/external/Makefile.mk` :facepalm: 

> I usually do `rm -f build/UEFIPAYLOAD.fd build/coreboot.rom` before make to force rebuild of EDK.

Same, I've lost too much work in payloads due to `make clean/distclean` before learning this trick.

SergiiDmytruk committed:
DasharoPayloadPkg/DasharoPayloadPkg.dsc: Enable FMP updates

This is using test certificate stored in this repository, but the root
include was copied to ease replacement with the production key.
Building additionally requires defining CAPSULE_FIRMWARE_GUID to
firmware GUID in string form.

FDF file is not updated because FMP driver will be embedded into update
capsule.

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
MdeModulePkg/Universal/CapsuleRuntimeDxe: Enable FUM on reset for capsule update

This should ensure that various protection mechanisms are off and won't
get in the way of firmware flashing.

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
DasharoModulePkg: Extract DasharoEnableFUM()

It will be used elsewhere.

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
DasharoPayloadPkg/SmmStore: Untie SmmStoreLib from runtime

The library was handling address switch event and updating pointers via
EfiConvertPointer() from UefiRuntimeLib which can't be used in a
non-runtime DXE.

Turn event handler into an exported function and make SmmStoreFvb invoke
that function in its handler of virtual address change.

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
DasharoPayloadPkg/SmmStore: Add API to read/write/erase any flash block

This allows reusing SMMSTORE protocol for the purpose of firmware
updates.

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
DasharoPayloadPkg/CbParseLib: Add parsing of string tags in CBMEM

Such strings include, for example, coreboot version information.

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
DasharoPayloadPkg/FmpDeviceSmmLib: Add for flashing via SMMSTOREv2

Part of functions of the library are left unimplemented (return
EFI_UNSUPPORTED) the rest use information about current firmware
obtained from CBMEM and SMMSTOREv2 to perform flashing.

Flashing is slightly optimized: FmpDeviceSetImageWithStatus() first
reads a block and checks that it differs from the new contents before
initiating erase and write.

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
MdeModulePkg/Universal/CapsuleRuntimeDxe: Reject capsules unless ME is HAP-disabled

HAP-disabled ME doesn't do anything, including writing to system flash,
which is what we need for a firmware update that relies on a warm reset.
coreboot assumes that HECI/soft-disabled state of ME isn't as good as
HMRFPO and switches to HMRFPO doing a global reset which loses in-RAM
capsules.

Checking variable's value should be enough, if somebody manually set it to
an invalid value, the update there will be a reboot without a capsule
update.  A more reliable solution would be to pass this information from
coreboot.

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
DasharoModulePkg/DasharoSystemFeaturesUiLib: Add UI note on ME vs. capsules

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk commented:
Force-pushed a rebase.

SergiiDmytruk commented:
Force-push:
 - rebase and resolve trivial conflicts
 - pick up root certificate public key from `DasharoPayloadPkg/CapsuleRootKey.inc` instead of deep inside `BaseTools` (make a copy of test key which can be updated with production key)
 - switch define from `CAPSULE_ENABLE` to `CAPSULE_SUPPORT` because that's what `UefiPayloadPkg` is using (despite tianocore's wiki)  and that's how changes were sent upstream

SergiiDmytruk commented:
Force-push:
 * rebase
 * [enable FUM for all persistent capsules](https://github.com/Dasharo/edk2/compare/1370be0e5fbbf72600642b4bd31f4f9626e343f2..f171cf764fda42cf6cc1af403db16e6b81df5fa8#diff-5489f2ff766e783be8bef65660b31a7305789423d16cdeb1809dd32791bf2125), not just those which initiate reset

SergiiDmytruk committed:
DasharoPayloadPkg/FmapLib: add

coreboot flash map data parsing library.

There is a small amount of code, so modify it to mostly look like EDK
code.

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
DasharoPayloadPkg/FmpDeviceSmmLib: read current flash at once

This is a preparation for moving data from current image.

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
DasharoPayloadPkg/FmpDeviceSmmLib: prepare for data migration

Introduce and integrate MergeFirmwareImages() which will perform the
migration.

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
DasharoPayloadPkg/FmpDeviceSmmLib: migrate SMMSTORE region

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
DasharoPayloadPkg/FmpDeviceSmmLib: migrate ROMHOLE

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
DasharoPayloadPkg/FmpDeviceSmmLib: migrate boot logo

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
DasharoPayloadPkg/CbfsLib: import

The code is imported in its essentially unchanged form (although large
chunks of unused code were removed).  There is a header which maps C
symbols to their equivalents in EDK2 to make the code work without
modifications.

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
MdeModulePkg/Logo: fallback to builtin logo on error

Returning an error from this function can disable useful functionality
like capsule updates because progress bar treats such errors as fatal...

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
DasharoPayloadPkg/FmpDeviceSmmLib: migrate SMBIOS data

The information is stored in CBFS files with known names.

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
DasharoPayloadPkg/EfiVarsLib: add

Simple EFI variables driver as a library.

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
DasharoPayloadPkg/FmpDeviceSmmLib: migrate EFI variables individually

Instead of copying whole SMMSTORE region.

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk commented:
Force-push: rebase onto the latest target branch.

krystian-hebel committed:
DasharoPayloadPkg/CapsuleSplashDxe: add driver for consistent progress bar

Position and size of progress bar produced by
DisplayUpdateProgressLibGraphics depends on the size of boot logo.
As the logo can be customized by users, a dummy logo with determinable
dimensions, depending on the size of current GOP mode, is created to
make progress bar appear always in the same place.

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

krystian-hebel committed:
CapsuleSplashDxe/SSFN: add scalable-font2 renderer

Project page: https://gitlab.com/bztsrc/scalable-font2

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

krystian-hebel committed:
CapsuleSplashDxe/CapsuleSplashDxe.c: print some warnings

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

krystian-hebel committed:
CapsuleSplashDxe/CapsuleSplashDxe.c: move text on the screen, ignore errors

Reasoning for ignoring errors is that this driver gives only
informational output, and failure to do so shouldn't abort
the update process.

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

SergiiDmytruk commented:
Force-push: rebase onto the latest target branch.

krystian-hebel committed:
DasharoPayloadPkg/BlSupportPei: fix PEI memory range discovery code

Previously, the code chose the highest non-reserved RAM region below
4G mark, regardless of its size. If the memory is fragmented, this
may result in failed memory allocation further down the line. Skip
regions smaller than PeiMemSize as defined in BlPeiEntryPoint.

In some cases, it isn't possible to differentiate between structures
in RAM (e.g. coreboot tables) and memory reserved by System Agent (e.g.
TSEG, GSM) based on memory region type. Both are reported as reserved
memory and they may even be combined into one region, but their caching
methods must be different. Because of that, MTRR setting code is no
longer called at this point. This depends on coreboot setting memory
caching attributes properly, which it already does.

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

krystian-hebel committed:
DasharoModulePkg/DasharoVariablesLib: add runtime attribute to FUM

This is required for initializing capsule update process from Linux's
capsule_loader.

It also fixes the possibility for putting the device in infinite FUM
loop by creating the variable with runtime access. SetVariable() in
WarnIfFirmwareUpdateMode() used to remove the variable failed because
of mismatched attributes.

According to UEFI specification,

> If a preexisting variable is rewritten with no access attributes
> specified, the variable will be deleted.

This is exactly what is needed. The other way to delete a variable is
to specify matching attributes and set the size of a variable as 0.

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

krystian-hebel commented:
https://github.com/Dasharo/edk2/pull/147/commits/5b3658c0503758f58861000a36251cd687a9d5a7 is pointed to by release tag for MSI and thus must be preserved either by keeping this branch or putting a tag on that commit.

SergiiDmytruk commented:
Pushed 2 tags:
 * https://github.com/Dasharo/edk2/releases/tag/msi_ms7d25_v1.1.4
 * https://github.com/Dasharo/edk2/releases/tag/msi_ms7e06_v0.9.2
